File: code-tiny/tools/struts/validation_parser.py
from __future__ import annotations

from pathlib import PurePath
from typing import List

from tools.struts.models import ValidationData, ValidationRule
from tools.struts.xml_utils import child_text, children, params, parse_xml


def _target_from_name(file_path: str) -> tuple[str, str]:
    stem = PurePath(file_path).name
    prefix = stem[: -len("-validation.xml")] if stem.endswith("-validation.xml") else PurePath(file_path).stem
    if "-" not in prefix:
        return prefix, ""
    target, method = prefix.rsplit("-", 1)
    return target, method


def parse_validation_xml_file(root: str, file_path: str) -> ValidationData:
    document, source, diagnostics = parse_xml(root, file_path, "struts.validation")
    if document is None:
        return ValidationData(diagnostics=diagnostics)

    target, method = _target_from_name(source.file_path)
    rules: List[ValidationRule] = []
    for element in document:
        tag = element.tag.rsplit("}", 1)[-1]
        if tag == "field":
            field_name = (element.get("name") or "").strip()
            validators = children(element, "field-validator")
        elif tag == "validator":
            field_name = ""
            validators = (element,)
        else:
            continue
        for validator in validators:
            message_element = next(children(validator, "message"), None)
            rules.append(
                ValidationRule(
                    target=target,
                    method=method,
                    validator_type=(validator.get("type") or "").strip(),
                    field_name=field_name,
                    message=child_text(validator, "message"),
                    message_key=(message_element.get("key") or "").strip() if message_element is not None else "",
                    params=params(validator),
                    source=source,
                )
            )
    return ValidationData(rules=tuple(rules), diagnostics=diagnostics)


__all__ = ["parse_validation_xml_file"]

File: rust/crates/analyzer-jvm-overlays/src/struts/validation_parser.rs
//! Port `tools/struts/validation_parser.py`.

use super::models::{ValidationData, ValidationRule};
use super::xml_utils::{child_text, children, params, parse_xml, text};

/// `_target_from_name` — tên file `Xyz-validation.xml` → (target, method).
fn target_from_name(file_path: &str) -> (String, String) {
    let stem = crate::pyutil::basename(file_path);
    let prefix = if let Some(stripped) = stem.strip_suffix("-validation.xml") {
        stripped.to_string()
    } else {
        stem.trim_end_matches(".xml").to_string()
    };
    // Python: PurePath(file_path).stem cho fallback — bỏ extension cuối.
    match prefix.rfind('-') {
        Some(index) if prefix.contains('-') => (prefix[..index].to_string(), prefix[index + 1..].to_string()),
        _ => (prefix.clone(), String::new()),
    }
}

pub fn parse_validation_xml_file(root: &str, file_path: &str) -> ValidationData {
    let (document, source, diagnostics) = parse_xml(root, file_path, "struts.validation");
    let Some(document) = document else {
        return ValidationData {
            diagnostics,
            ..Default::default()
        };
    };

    let (target, method) = target_from_name(&source.file_path);
    let mut rules: Vec<ValidationRule> = Vec::new();
    for element in document.elements() {
        let tag = element.tag.rsplit('}').next().unwrap_or(&element.tag).to_string();
        let (field_name, validators): (String, Vec<&super::xml_dom::XmlElement>) = if tag == "field" {
            let name = element.attr("name").unwrap_or("").trim().to_string();
            (name, children(element, "field-validator"))
        } else if tag == "validator" {
            (String::new(), vec![element])
        } else {
            continue;
        };
        for validator in validators {
            let message_element = children(validator, "message").into_iter().next();
            rules.push(ValidationRule {
                target: target.clone(),
                method: method.clone(),
                validator_type: validator.attr("type").unwrap_or("").trim().to_string(),
                field_name: field_name.clone(),
                message: child_text(validator, "message", ""),
                message_key: message_element
                    .and_then(|element| element.attr("key"))
                    .unwrap_or("")
                    .trim()
                    .to_string(),
                params: params(validator),
                source: source.clone(),
            });
        }
    }
    ValidationData {
        rules,
        diagnostics,
    }
}

/// `text` re-export (unused).
#[allow(dead_code)]
fn unused(element: Option<&super::xml_dom::XmlElement>) -> String {
    text(element)
}
